Fix inconsistant logical index behavior

N5N3 opened the pull request:
1. If we use `BitArray`/`Array{Bool}` to index, `to_indices` has an optimiztion for linear-iteratable case. 
But the corresponding check is not correct. IIUC, this optimization is legal only when the Boolen array is the only index provided. 
The first commit fix it and widen this optimization to all Boolen array.
Before this PR
```julia
julia> A = rand(2,3,4); I = rand(Bool,3,4);

julia> A[1,I] == A[1,view(I,:,:)]
ERROR: BoundsError: attempt to access 2×3×4 Array{Float64, 3} at index [1, 3×4 Matrix{Bool}]
```
After
```julia
julia> A = rand(2,3,4); I = rand(Bool,3,4);

julia> A[1,I] == A[1,view(I,:,:)]
true
```

2. On master, if the index/array has singleton trailing dimension, boundcheck of logical index show different behavior depending on the number of indexes provided.
If there's only one index variable, singleton dimension wil not be ignored. The second commit fix it. (close #45867)

N5N3 committed:
Enabled linear optimization for logical indexes only if they are the only index variable.

`zeros(2,3,4)[2, 1]` is always invalid, thus this optimization is illegal for trailing dimension.

N5N3 committed:
Some doc correction.

N5N3 commented on doc/src/manual/arrays.md:
```diff
@@ -793,38 +793,46 @@ Indexing by a boolean vector `B` is effectively the same as indexing by the
-array it indexes into. It is generally more efficient to use boolean arrays as
-indices directly instead of first calling [`findall`](@ref).
+must be a array of the same shape as the dimension(s) it indexes into, or it 
+must be the only index provided and match the shape of the one-dimensional 
+reshaped view of the array it indexes into. It is generally more efficient 
```
This is still not entirely accurate as `CartesianIndex{0}` could always be ignored:
```julia
julia> a = randn(2,2); I = rand(Bool, 4);

julia> a[CartesianIndex(),I,CartesianIndex()] == a[I] # Both on LTS and master.
true
```
But I wonder do we need to mention it here.

ViralBShah commented:
Who could review this? @StefanKarpinski?

N5N3 committed:
Partially revert #45672

This no more influence inference on master.

StefanKarpinski commented:
@mbauman would be a good person to review when he has the chance.

mbauman commented on base/multidimensional.jl:
```diff
@@ -860,21 +834,30 @@ ensure_indexable(I::Tuple{}) = ()
+_cutdim(inds, I::AbstractArray{Bool}) = IteratorsMD.split(inds, Val(ndims(I)))[2]
+# As an optimization, we allow the only `AbstractArray{Bool}` to be linear-iterated
+@inline to_indices(A, I::Tuple{AbstractArray{Bool}}) = (_maybe_linear_logical_index(A, I[1]),)
+# Or if it's the first n-d (n > 0) index.
```
I think this is an oblique-enough case that we shouldn't worry about using linear indexing here.  It's far safer to just use the cartesian index and give up on the optimization.

mbauman commented on base/multidimensional.jl:
```diff
@@ -860,21 +834,30 @@ ensure_indexable(I::Tuple{}) = ()
-_cutdim(inds, I1::CartesianIndex) = IteratorsMD.split(inds, Val(length(I1)))[2]
 # For arrays of CartesianIndex, we just skip the appropriate number of inds
-_cutdim(inds, I1::AbstractArray{CartesianIndex{N}}) where {N} = IteratorsMD.split(inds, Val(N))[2]
+_cutdim(inds, ::AbstractArray{CartesianIndex{N}}) where {N} = IteratorsMD.split(inds, Val(N))[2]
```
It'd be nice to coalesce this around `index_ndims`, too.  Obviously that's not new to this PR, but would be great to cleanup at the same time.

N5N3 replied:
Since the inference of indirect rescusion get improved, I think we can revert these change from #45672.

mbauman commented on base/abstractarray.jl:
```diff
@@ -731,19 +730,23 @@ of `IA`.
 checkbounds_indices(::Type{Bool}, ::Tuple{}, ::Tuple{}) = true
 
-throw_boundserror(A, I) = (@noinline; throw(BoundsError(A, I)))
+ndims_cartesian_index(::Any) = nothing # indexes are non-cartesian by default
```
Instead of introducing yet another way of getting this information, it'd be good to lean on the existing `index_ndims` functionality.

N5N3 replied:
The problem here is we need to check whether the input supports linear indexing.

`checkbounds(zeros(2, 2), CartesianIndex(1))` throws, but `checkbounds(zeros(2, 2), 1)` is OK.
`index_ndims` can't distinguish them. Unfortunately, we'd better not change it's behavior, as it seems to be used widely outside Base.

mbauman replied:
Aha, that makes way more sense to me now. OK, perhaps this should be _supports_linear_indexing, returning a Bool? I don't think we need the counting-dimensions behaviors to be mixed up in this, do we?

mbauman reviewed (commented):
Sorry to leave this languishing so long — I'm pretty sure I've tried reviewing it long ago but it's pretty complicated... and then I forgot about it.  See my comments, I think it'd be good to simplify/coalesce some of this logic while we're here instead of making it yet more complicated.  It feels like this is missing a primitive — maybe an internal function  like `_peel_axes_for_index` (that would effectively do a `peel` for the appropriate number of axes) would help simplify many of these functions?  I've been trying to get such a refactor written but I've simply not had time to see it to the end yet.

N5N3 committed:
`checkbounds` fix and some style clean.

Implement `checkbounds` using `to_indices` style.
The behavior of non-logical doesn't change.
While for `AbstractArray{Bool,N}`s, they would behaves like `Vector{CartesianIndex{N}}` after this commit, if its shape matches the indexed dimensions.
Thus the trailing singleton dimensions in boolean array would be ingnored only when we have comsumed all of the src dimensions.

The test change includes:
1. disallow `zeros(2,2)[trues(1,2), 1]`
2. allow `zeros(2,2)[trues(2,2,1)]`

mbauman commented on base/multidimensional.jl:
```diff
@@ -848,11 +820,29 @@ end
+end
+checkindex(::Type{Bool}, inds::AbstractUnitRange, I::AbstractVector{Bool}) = axes1(I) == inds
+checkindex(::Type{Bool}, inds::AbstractUnitRange, I::AbstractRange{Bool}) = axes1(I) == inds
+checkindex(::Type{Bool}, inds::Tuple, I::AbstractArray{Bool}) = _check_boolen_axes(inds, axes(I))
+_check_boolen_axes(inds::Tuple, axes::Tuple) = (inds[1] == axes[1]) & _check_boolen_axes(tail(inds), tail(axes))
+_check_boolen_axes(::Tuple{}, axes::Tuple) = all(==(OneTo(1)), axes)
```
```suggestion
checkindex(::Type{Bool}, inds::Tuple, I::AbstractArray{Bool}) = _check_boolean_axes(inds, axes(I))
_check_boolean_axes(inds::Tuple, axes::Tuple) = (inds[1] == axes[1]) & _check_boolean_axes(tail(inds), tail(axes))
_check_boolean_axes(::Tuple{}, axes::Tuple) = all(==(OneTo(1)), axes)
```

mbauman committed:
minor spelling fix

N5N3 committed:
Update base/abstractarray.jl

Co-authored-by: Matt Bauman <mbauman@gmail.com>

N5N3 committed:
Merge branch 'master' into logicalindex

maleadt commented:
In order to adapt CUDA.jl to this, which requires iteratable logical indices too, is this the correct approach?

```julia
# we cannot use Base.LogicalIndex, which does not support indexing but requires iteration.
Base.to_index(::CuArray, I::AbstractArray{Bool}) = findall(I)
if VERSION >= v"1.11.0-DEV.1157"
    # for this PR
    Base.to_indices(A::CuArray, I::Tuple{AbstractArray{Bool}}) = (Base.to_index(A, I[1]),)
else
    # what we used to have
    Base.to_indices(A::CuArray, inds,
                    I::Tuple{Union{Array{Bool,N}, BitArray{N}}}) where {N} =
        (Base.to_index(A, I[1]),)
end
```
